### DISCUSSION TITLE
Updating a line

### DISCUSSION THREAD

**Discussion by XARP616** [0x20F/paris#54] [Q&A]
I'm having a read through the documentation and I found a super cool method which displays a [loading](https://docs.rs/paris/latest/paris/struct.Logger.html#method.loading) animation. I was wondering if it was possible to update such message on the fly. 

I will explain myself: let us say that we have a super heavy loading method which wants to output the percentage of work done in different steps and, instead of printing new lines, it wants to update the previous line that was outputted with new information/percentage.

The way I figured out is to stop and reload the animation:
```rust
logger.loading("0% bla bla bla");
sleep(Duration::from_millis(1000));
logger.done(); // If I don't do this next loading bugs

logger.loading("50% Some text over here");
sleep(Duration::from_millis(1000));

logger.log("Done");
```
Tho it's not the cleanest, is not the terriblest.

So, coming back to the question: is curretly possible to update a line (be it a loading or not) with library methods?

**Comment by 0x20F** [comment]
**Quick answer**: It's not possible.

However, I just pushed a quick fix for the bug you encountered when calling `.loading()` two times in a row. PR #55, should be merged soon, and it should be fixed for your specific use case.

Keeping with the theme though. Just for future reference, what would feel like a clean, intuitive API that would allow for clearing every log line if needed? Would having to call `logger.clearable()` before printing a line that will eventually get replaced be a good option?

**Comment by XARP616** [comment]
Thank you for fixing it! 

About the API, good question. Something like `logger.clearable()` or `logger.updateable()` sounds allright. Then, calling another method for updating it such as `logger.update()`, or maybe `logger.update_loading()` to be more clear that it only works for loading messages.

A `logger.clear_loading()` may be handy as well, tho it doesn't make much use if you plan writing afterwards but, who knows?

What I was thinking may look more or less like this:
```rust
logger.loading("0% bla bla bla").clearable(); 
sleep(Duration::from_millis(1000));
// logger.done(); // bug fixed! :)

logger.update("50% Some text over here");
sleep(Duration::from_millis(1000));

logger.done("Finished!");
```

If you are looking for more ideas, maybe including a loading bar similar to Python's pip or apt's would be nice, where you can set the progress along your code:
```rust
logger.loading_bar("Creating temp folders")
// do stuff
logger.set_progress("Creating files", 10); // message, %
// do more stuff
logger.set_progress("Analyzing for malware", 50);
[...]
logger.done("Finished! 1305 viruses found");
```

Hope you find it useful. I love how the project looks and best wishes ;b
